pkg/assets/builder.go

```diff
@@ -127,32 +142,123 @@ func (a *AssetBuilder) RemapImage(image string) (string, error) {
+
+	a.FileAssets = append(a.FileAssets, fileAsset)
+	glog.V(8).Infof("adding file: %+v", fileAsset)
+
+	return fileAsset.FileURL, h, nil
+}
+
+// TODO - remove this method as CNI does now have a SHA file
+
+// RemapFileAndSHAValue is used exclusively to remap the cni tarball, as the tarball does not have a sha file in object storage.
+func (a *AssetBuilder) RemapFileAndSHAValue(fileURL *url.URL, shaValue string) (*url.URL, error) {
+	if fileURL == nil {
+		return nil, fmt.Errorf("unable to remap an empty URL")
 	}
 
 	fileAsset := &FileAsset{
-		File:              file,
-		CanonicalLocation: file,
+		FileURL:  fileURL,
+		SHAValue: shaValue,
 	}
 
 	if a.AssetsLocation != nil && a.AssetsLocation.FileRepository != nil {
-		fileURL, err := url.Parse(file)
+		fileAsset.CanonicalFileURL = fileURL
+
+		normalizedFile, err := a.normalizeURL(fileURL)
 		if err != nil {
-			return "", fmt.Errorf("unable to parse file url %q: %v", file, err)
+			return nil, fmt.Errorf("unable to parse file url %q: %v", normalizedFile, err)
 		}
 
-		fileRepo := strings.TrimSuffix(*a.AssetsLocation.FileRepository, "/")
-		fileAsset.File = fileRepo + fileURL.Path
+		fileAsset.FileURL = normalizedFile
+		glog.V(4).Infof("adding remapped file: %q", fileAsset.FileURL.String())
 	}
 
 	a.FileAssets = append(a.FileAssets, fileAsset)
 
-	return fileAsset.File, nil
+	return fileAsset.FileURL, nil
+}
+
+// FindHash returns the hash value of a FileAsset.
+func (a *AssetBuilder) findHash(file *FileAsset) (*hashing.Hash, error) {
+	u := file.FileURL
+
+	// If the phase is "assets" we use the CanonicalFileURL,
+	// but during other phases we use the hash from the FileRepository or the base kops path.
+	// We do not want to just test for CanonicalFileURL as it is defined in
+	// other phases, but is not used to test for the SHA.
+	// This prevents a chicken and egg problem where the file is not yet in the FileRepository.
+	// Summary
+	// assets phase -> get the sha file from the source / CanonicalFileURL
+	// any other phase -> get the sha file from the kops base location or the FileRepository
+	if a.Phase == "assets" && file.CanonicalFileURL != nil {
+		u = file.CanonicalFileURL
+	}
+
+	if u == nil {
+		return nil, fmt.Errorf("file url is not defined")
+	}
+
+	// TODO This is old code that can be removed from the for loop.
+	// TODO This was in place in case we have different file extensions for sha files.
+	for _, ext := range []string{".sha1"} {
+		hashURL := u.String() + ext
+		b, err := vfs.Context.ReadFile(hashURL)
+		if err != nil {
+			glog.Infof("error reading hash file %q: %v", hashURL, err)
+			continue
+		}
+		hashString := strings.TrimSpace(string(b))
+		glog.V(2).Infof("Found hash %q for %q", hashString, u)
+
+		return hashing.FromString(hashString)
+	}
+
+	if a.AssetsLocation != nil && a.AssetsLocation.FileRepository != nil {
+		glog.Errorf("you may have not staged you files correctly, please execute kops update cluster using the asessts phase")
```

Comment:
typo: "your files correctly"

typo: "assets phase"

Should this be a `return nil, fmt.Errorf(` ?